osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/factories/InvocationCounterManager.java

```diff
@@ -0,0 +1,82 @@
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.repositories.DlmsDeviceRepository;
+import org.opensmartgridplatform.shared.exceptionhandling.FunctionalException;
+import org.opensmartgridplatform.shared.exceptionhandling.OsgpException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Object that manages initializing and resetting the value of the invocationCounter property of a device.
+ */
+@Component
+public class InvocationCounterManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InvocationCounterManager.class);
+    private static final AttributeAddress ATTRIBUTE_ADDRESS_INVOCATION_COUNTER_VALUE = new AttributeAddress(1,
+            new ObisCode(new byte[] { 0, 0, 43, 1, 0, -1 }), 2);
+
+    private final DlmsConnectionFactory connectionFactory;
+    private final DlmsHelperService dlmsHelper;
+    private final DlmsDeviceRepository deviceRepository;
+
+    @Autowired
+    public InvocationCounterManager(final DlmsConnectionFactory connectionFactory, final DlmsHelperService dlmsHelper,
+            final DlmsDeviceRepository deviceRepository) {
+        this.connectionFactory = connectionFactory;
+        this.dlmsHelper = dlmsHelper;
+        this.deviceRepository = deviceRepository;
+    }
+
+    /**
+     * Updates the device instance with the invocation counter value on the actual device.
+     */
+    public void initializeInvocationCounter(final DlmsDevice device) throws OsgpException {
+        if (this.invocationCounterIsStoredOnDevice(device)) {
+            this.initializeWithInvocationCounterStoredOnDevice(device);
+        } else {
+            // Value of invocation counter is ignored on these devices.
+            device.setInvocationCounter(0);
+        }
+        this.deviceRepository.save(device);
+    }
+
+    private void initializeWithInvocationCounterStoredOnDevice(final DlmsDevice device) throws OsgpException {
+        try (final DlmsConnectionManager connectionManager = this.connectionFactory
+                .getPublicClientConnection(device, null)) {
+            device.setInvocationCounter(this.getInvocationCounter(connectionManager));
+            LOGGER.info(
+                    "Property invocationCounter of device {} initialized to the value of the invocation counter "
+                            + "stored on the device: {}", device.getDeviceIdentification(),
+                    device.getInvocationCounter());
+        }
+    }
+
+    private boolean invocationCounterIsStoredOnDevice(final DlmsDevice device) {
+        return "SMR".equals(device.getProtocol());
+    }
+
+    private int getInvocationCounter(final DlmsConnectionManager connectionManager) throws FunctionalException {
+        return ((Number) this.dlmsHelper
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unnecessary cast to "Number". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1905)